### ISSUE TITLE
Latest version 2.4.0 not available for Mac M1, python 3.10 on pypi

### ISSUE THREAD

**Issue by matthiaskoenig** [sys-bio/roadrunner#1147]
Hi all,

The latest versions are not available for 
Mac, Apple M1, Ventura 13.5.2, python 3.10
from pypi. Only libroadrunner-2.2.0 is available.

Somehow the latest releases were not uploaded for these platforms. 
I have some students using the systems and we want to use the latest version.

Best Matthias

**Comment by matthiaskoenig** [comment]
Here the release files for 2.4.0
https://pypi.org/project/libroadrunner/2.4.0/#files
libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl
libroadrunner-2.4.0-cp310-cp310-macosx_11_0_x86_64.whl

vs
https://pypi.org/project/libroadrunner/2.2.2/#files
libroadrunner-2.2.2-cp310-cp310-macosx_12_0_universal2.whl
libroadrunner-2.2.2-cp310-cp310-macosx_10_15_x86_64.whl

To work on the M1 architecture you probably also require the universal architecture, i.e., there should probably be pypi files for arm64, x86_64 and universal.

@MadNietzsche please follow the issue to check when the release will be available

**Comment by luciansmith** [comment]
I'm not sure what to do here.  The 2.2.2 and the 2.4.0 files were both created on the same M1 laptop, though I upgraded it between releases to 13.0 instead of 12.  But the reason I switched to arm64 from universal is that the so-called 'universal' binaries didn't actually work on non-M1 machines, I think because the system libraries weren't cross-compiled.   I don't have 'universal' dependencies, and I'm pretty sure I never did!

Is 13.0 not compatible with 13.5?  I can upgrade the laptop and compile the wheels again.  Or maybe it would just work if you renamed the wheel?

**Comment by matthiaskoenig** [comment]
@MadNietzsche Could you try to install the 2.4.0 wheel manually. We should then figure out why pip cannot resolve the pypi package as an install candidate on your machine. Perhaps there are some pip options to force the install. Please update pip on your machine/virtual env and try again. Also try the manual install until our next meeting.

**Comment by matthiaskoenig** [comment]
The last three components of the filename before the extension are called “compatibility tags.” The compatibility tags express the package’s basic interpreter requirements and are detailed in [PEP 425](https://peps.python.org/pep-0425/).
The one which is relevant here is the platform tag. So most likely the platform is not reported correctly for the M1 machine. 
There is an option in pip to set the platform manually
```
--platform <platform>       Only use wheels compatible with <platform>. Defaults to the platform of the running system. Use this option multiple times to
                              specify multiple platforms supported by the target interpreter.
```
So the following should likely work on the system
```
pip install libroadrunner>=2.4.0 --platform macosx_13_0_arm64
```
it is clearly a MacOS 13 with an arm64, so that should fix it. The interesting part is why the platform your system is reporting is not compatible to these platform tag.
Your platform tag is simply `distutils.util.get_platform()` with all hyphens - and periods . replaced with underscore _.
@MadNietzsche please try if the above solution works for you and would be great to know your platform tag to figure out what is going on.

**Comment by MadNietzsche** [comment]
Hi everyone, 

I wasn't able to do `pip install libroadrunner>=2.4.0 --platform macosx_13_0_arm64` . The message I get is 
```
zsh: 2.4.0 not found
```
Here are my system profile, python version and pip list
```
> system_profiler SPSoftwareDataType SPHardwareDataType
Software:

    System Software Overview:

      System Version: macOS 13.5.2 (22G91)
      Kernel Version: Darwin 22.6.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: de-aaguilar
      User Name: Antonio Aquiles Alvarez Aguilar (aaguilar)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 10 days, 12 hours, 3 minutes

Hardware:

    Hardware Overview:

      Model Name: MacBook Air
      Model Identifier: MacBookAir10,1
      Model Number: Z124000JHD/A
      Chip: Apple M1
      Total Number of Cores: 8 (4 performance and 4 efficiency)
      Memory: 16 GB
      System Firmware Version: 8422.141.2
      OS Loader Version: 8422.141.2
      Serial Number (system): FVFHW11PQ6LR
      Hardware UUID: 6AA14254-15DB-538D-BCDD-0233A2D61C75
      Provisioning UDID: 00008103-000121500253001E
      Activation Lock Status: Disabled

> python -V
Python 3.10.13

> pip list
Package                   Version
------------------------- ------------
annotated-types           0.6.0
antimony                  2.13.2
anyio                     3.7.1
appnope                   0.1.2
argon2-cffi               21.3.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
arviz                     0.16.1
asttokens                 2.0.5
async-lru                 2.0.4
attrs                     23.1.0
Babel                     2.13.0
backcall                  0.2.0
beautifulsoup4            4.12.2
bleach                    4.1.0
boltons                   23.0.0
brotlipy                  0.7.0
cachetools                5.3.2
certifi                   2023.7.22
cffi                      1.15.1
chardet                   5.2.0
charset-normalizer        2.0.4
click                     8.1.7
cloudpickle               3.0.0
comm                      0.1.2
conda                     23.9.0
conda-package-handling    2.2.0
conda_package_streaming   0.9.0
cons                      0.4.6
contourpy                 1.1.1
cryptography              41.0.3
cycler                    0.12.1
debugpy                   1.6.7
decorator                 5.1.1
defusedxml                0.7.1
depinfo                   2.2.0
elementpath               4.1.5
entrypoints               0.4
et-xmlfile                1.1.0
etuples                   0.3.9
exceptiongroup            1.0.4
executing                 0.8.3
fastapi                   0.103.2
fastjsonschema            2.16.2
fastobo                   0.12.2
fastprogress              1.0.3
filelock                  3.13.0
fonttools                 4.43.1
fqdn                      1.5.1
greenlet                  2.0.2
h11                       0.14.0
h5netcdf                  1.2.0
h5py                      3.10.0
idna                      3.4
igraph                    0.11.2
ipykernel                 6.25.0
ipython                   8.15.0
ipython-genutils          0.2.0
isodate                   0.6.1
isoduration               20.11.0
jedi                      0.18.1
Jinja2                    3.1.2
json5                     0.9.14
jsonpatch                 1.32
jsonpointer               2.1
jsonschema                4.17.3
jsonschema-specifications 2023.7.1
jupyter_client            7.4.9
jupyter_core              5.3.0
jupyter-events            0.8.0
jupyter-lsp               2.2.0
jupyter-server            1.23.4
jupyter_server_terminals  0.4.4
jupyterlab                4.0.7
jupyterlab-pygments       0.1.2
jupyterlab_server         2.25.0
kiwisolver                1.4.5
libroadrunner             2.2.2
llvmlite                  0.41.1
logical-unification       0.4.6
lxml                      4.9.3
markdown-it-py            3.0.0
MarkupSafe                2.1.1
matplotlib                3.8.0
matplotlib-inline         0.1.6
mdurl                     0.1.2
miniKanren                1.0.3
mistune                   0.8.4
multipledispatch          1.0.0
nb-conda-kernels          2.3.1
nbclassic                 0.5.5
nbclient                  0.5.13
nbconvert                 6.5.4
nbformat                  5.9.2
nest-asyncio              1.5.6
networkx                  3.1
notebook                  6.5.4
notebook_shim             0.2.2
numba                     0.58.1
numpy                     1.25.2
openpyxl                  3.1.2
overrides                 7.4.0
packaging                 23.1
pandas                    2.1.1
pandocfilters             1.5.0
parso                     0.8.3
pexpect                   4.8.0
pickleshare               0.7.5
Pillow                    10.1.0
Pint                      0.22
pip                       23.3
platformdirs              3.10.0
playwright                1.38.0
pluggy                    1.0.0
prometheus-client         0.14.1
prompt-toolkit            3.0.36
pronto                    2.5.5
psutil                    5.9.0
ptyprocess                0.7.0
pure-eval                 0.2.2
py2cytoscape              0.7.1
pycosat                   0.6.6
pycparser                 2.21
pydantic                  2.4.2
pydantic_core             2.10.1
pydot                     1.4.2
pydotplus                 2.0.2
pyee                      9.0.4
Pygments                  2.15.1
pymc                      5.1.2
pymetadata                0.4.0
pyOpenSSL                 23.2.0
pyparsing                 3.1.1
pyrsistent                0.18.0
PySocks                   1.7.1
pytensor                  2.10.1
python-dateutil           2.8.2
python-igraph             0.11.2
python-json-logger        2.0.7
python-libsbml            5.20.2
python-multipart          0.0.6
pytz                      2023.3.post1
PyYAML                    6.0.1
pyzmq                     23.2.0
referencing               0.30.2
requests                  2.31.0
requests-file             1.5.1
requests-toolbelt         1.0.0
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      13.6.0
rpds-py                   0.10.6
ruamel.yaml               0.17.21
ruamel.yaml.clib          0.2.6
sbmlutils                 0.8.7
scipy                     1.11.3
Send2Trash                1.8.0
setuptools                68.0.0
six                       1.16.0
sniffio                   1.2.0
soupsieve                 2.5
stack-data                0.2.0
starlette                 0.27.0
tabulate                  0.9.0
terminado                 0.17.1
texttable                 1.7.0
tinycss2                  1.2.1
tomli                     2.0.1
toolz                     0.12.0
tornado                   6.3.3
tqdm                      4.65.0
traitlets                 5.11.2
truststore                0.8.0
types-python-dateutil     2.8.19.14
typing_extensions         4.7.1
tzdata                    2023.3
uri-template              1.3.0
urllib3                   1.26.16
uuid                      1.30
uvicorn                   0.23.2
wcwidth                   0.2.5
webcolors                 1.13
webencodings              0.5.1
websocket-client          0.58.0
wheel                     0.41.2
xarray                    2023.10.1
xarray-einstats           0.6.0
xmlschema                 2.5.0
xmltodict                 0.13.0
zeep                      4.2.1
zstandard                 0.19.0


```

**Comment by MadNietzsche** [comment]
Hi everyone, 

I wasn't able to do `pip install libroadrunner>=2.4.0 --platform macosx_13_0_arm64` . The message I get is 
```
zsh: 2.4.0 not found
```
Here are my system profile (Hardware IDs deleted), python version and pip list
```
> system_profiler SPSoftwareDataType SPHardwareDataType
Software:

    System Software Overview:

      System Version: macOS 13.5.2 (22G91)
      Kernel Version: Darwin 22.6.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: de-aaguilar
      User Name: Antonio Aquiles Alvarez Aguilar (aaguilar)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 10 days, 12 hours, 3 minutes

Hardware:

    Hardware Overview:

      Model Name: MacBook Air
      Model Identifier: MacBookAir10,1
      Model Number: Z124000JHD/A
      Chip: Apple M1
      Total Number of Cores: 8 (4 performance and 4 efficiency)
      Memory: 16 GB
      System Firmware Version: 8422.141.2
      OS Loader Version: 8422.141.2
      Activation Lock Status: Disabled

> python -V
Python 3.10.13

> pip list
Package                   Version
------------------------- ------------
annotated-types           0.6.0
antimony                  2.13.2
anyio                     3.7.1
appnope                   0.1.2
argon2-cffi               21.3.0
argon2-cffi-bindings      21.2.0
arrow                     1.3.0
arviz                     0.16.1
asttokens                 2.0.5
async-lru                 2.0.4
attrs                     23.1.0
Babel                     2.13.0
backcall                  0.2.0
beautifulsoup4            4.12.2
bleach                    4.1.0
boltons                   23.0.0
brotlipy                  0.7.0
cachetools                5.3.2
certifi                   2023.7.22
cffi                      1.15.1
chardet                   5.2.0
charset-normalizer        2.0.4
click                     8.1.7
cloudpickle               3.0.0
comm                      0.1.2
conda                     23.9.0
conda-package-handling    2.2.0
conda_package_streaming   0.9.0
cons                      0.4.6
contourpy                 1.1.1
cryptography              41.0.3
cycler                    0.12.1
debugpy                   1.6.7
decorator                 5.1.1
defusedxml                0.7.1
depinfo                   2.2.0
elementpath               4.1.5
entrypoints               0.4
et-xmlfile                1.1.0
etuples                   0.3.9
exceptiongroup            1.0.4
executing                 0.8.3
fastapi                   0.103.2
fastjsonschema            2.16.2
fastobo                   0.12.2
fastprogress              1.0.3
filelock                  3.13.0
fonttools                 4.43.1
fqdn                      1.5.1
greenlet                  2.0.2
h11                       0.14.0
h5netcdf                  1.2.0
h5py                      3.10.0
idna                      3.4
igraph                    0.11.2
ipykernel                 6.25.0
ipython                   8.15.0
ipython-genutils          0.2.0
isodate                   0.6.1
isoduration               20.11.0
jedi                      0.18.1
Jinja2                    3.1.2
json5                     0.9.14
jsonpatch                 1.32
jsonpointer               2.1
jsonschema                4.17.3
jsonschema-specifications 2023.7.1
jupyter_client            7.4.9
jupyter_core              5.3.0
jupyter-events            0.8.0
jupyter-lsp               2.2.0
jupyter-server            1.23.4
jupyter_server_terminals  0.4.4
jupyterlab                4.0.7
jupyterlab-pygments       0.1.2
jupyterlab_server         2.25.0
kiwisolver                1.4.5
libroadrunner             2.2.2
llvmlite                  0.41.1
logical-unification       0.4.6
lxml                      4.9.3
markdown-it-py            3.0.0
MarkupSafe                2.1.1
matplotlib                3.8.0
matplotlib-inline         0.1.6
mdurl                     0.1.2
miniKanren                1.0.3
mistune                   0.8.4
multipledispatch          1.0.0
nb-conda-kernels          2.3.1
nbclassic                 0.5.5
nbclient                  0.5.13
nbconvert                 6.5.4
nbformat                  5.9.2
nest-asyncio              1.5.6
networkx                  3.1
notebook                  6.5.4
notebook_shim             0.2.2
numba                     0.58.1
numpy                     1.25.2
openpyxl                  3.1.2
overrides                 7.4.0
packaging                 23.1
pandas                    2.1.1
pandocfilters             1.5.0
parso                     0.8.3
pexpect                   4.8.0
pickleshare               0.7.5
Pillow                    10.1.0
Pint                      0.22
pip                       23.3
platformdirs              3.10.0
playwright                1.38.0
pluggy                    1.0.0
prometheus-client         0.14.1
prompt-toolkit            3.0.36
pronto                    2.5.5
psutil                    5.9.0
ptyprocess                0.7.0
pure-eval                 0.2.2
py2cytoscape              0.7.1
pycosat                   0.6.6
pycparser                 2.21
pydantic                  2.4.2
pydantic_core             2.10.1
pydot                     1.4.2
pydotplus                 2.0.2
pyee                      9.0.4
Pygments                  2.15.1
pymc                      5.1.2
pymetadata                0.4.0
pyOpenSSL                 23.2.0
pyparsing                 3.1.1
pyrsistent                0.18.0
PySocks                   1.7.1
pytensor                  2.10.1
python-dateutil           2.8.2
python-igraph             0.11.2
python-json-logger        2.0.7
python-libsbml            5.20.2
python-multipart          0.0.6
pytz                      2023.3.post1
PyYAML                    6.0.1
pyzmq                     23.2.0
referencing               0.30.2
requests                  2.31.0
requests-file             1.5.1
requests-toolbelt         1.0.0
rfc3339-validator         0.1.4
rfc3986-validator         0.1.1
rich                      13.6.0
rpds-py                   0.10.6
ruamel.yaml               0.17.21
ruamel.yaml.clib          0.2.6
sbmlutils                 0.8.7
scipy                     1.11.3
Send2Trash                1.8.0
setuptools                68.0.0
six                       1.16.0
sniffio                   1.2.0
soupsieve                 2.5
stack-data                0.2.0
starlette                 0.27.0
tabulate                  0.9.0
terminado                 0.17.1
texttable                 1.7.0
tinycss2                  1.2.1
tomli                     2.0.1
toolz                     0.12.0
tornado                   6.3.3
tqdm                      4.65.0
traitlets                 5.11.2
truststore                0.8.0
types-python-dateutil     2.8.19.14
typing_extensions         4.7.1
tzdata                    2023.3
uri-template              1.3.0
urllib3                   1.26.16
uuid                      1.30
uvicorn                   0.23.2
wcwidth                   0.2.5
webcolors                 1.13
webencodings              0.5.1
websocket-client          0.58.0
wheel                     0.41.2
xarray                    2023.10.1
xarray-einstats           0.6.0
xmlschema                 2.5.0
xmltodict                 0.13.0
zeep                      4.2.1
zstandard                 0.19.0


```

**Comment by luciansmith** [comment]
OK, try this:

download https://files.pythonhosted.org/packages/5f/0d/7dfd5ec11df45a11b20b912f733e605ca9a808eac86b98ed295ddcc4a706/libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl

pip install libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl

Tell us what the error message is.

Then try renaming the libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl file to libroadrunner-2.4.0-cp310-cp310-universal2.whl

and then:

pip install libroadrunner-2.4.0-cp310-cp310-universal2.whl

If that installs, the next thing to do would be to see if it runs:

pip install tellurium
python3.10
> import tellurium as te
> te.loada("species a=3")
> te.simulate()
> te.plot()

**Comment by MadNietzsche** [comment]
In both cases the error is the same:

**Comment by MadNietzsche** [comment]
In both cases the error is the same:

```
> pip install libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl
ERROR: libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl is not a supported wheel on this platform.

> mv libroadrunner-2.4.0-cp310-cp310-macosx_13_0_arm64.whl libroadrunner-2.4.0-cp310-cp310-universal2.whl

> pip install libroadrunner-2.4.0-cp310-cp310-universal2.whl
ERROR: libroadrunner-2.4.0-cp310-cp310-universal2.whl is not a supported wheel on this platform

```

**Comment by luciansmith** [comment]
OK, more diagnoses: Can you try, from python:

```python
from distutils import util
util.get_platform()
```
and report what it says?  And then maybe try renaming the end of the wheel file to that to see if it happens to work anyway?

Sorry about the weirdness here; this is very confusing.

**Comment by matthiaskoenig** [comment]
@MadNietzsche Could you give some feedback on the issue to get this running on your system?

**Comment by luciansmith** [comment]
We're well past v2.4.0, so I'm closing this for now.  If the latest versions aren't available soon, do open a new issue.

(The M1 version 2.6.0 should be available, but not the 2.7.0 yet; working on a new build system that should (finally!) include it.)
